pkg/postage/service.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package postage
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/storage"
+)
+
+const (
+	postagePrefix = "postage"
+)
+
+var (
+	// ErrNotFound is the error returned when issuer with given batch ID does not exist.
+	ErrNotFound = errors.New("not found")
+)
+
+// Service handles postage batches
+// stores the active batches.
+type Service struct {
+	lock    sync.Mutex
+	store   storage.StateStorer
+	chainID int64
+	issuers []*StampIssuer
+}
+
+// NewService constructs a new Service.
+func NewService(store storage.StateStorer, chainID int64) *Service {
+	return &Service{
+		store:   store,
+		chainID: chainID,
+	}
+}
+
+// Add adds a stamp issuer to the active issuers.
+func (ps *Service) Add(st *StampIssuer) {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+	ps.issuers = append(ps.issuers, st)
+}
+
+// StampIssuers returns the currently active stamp issuers.
+func (ps *Service) StampIssuers() []*StampIssuer {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+	return ps.issuers
+}
+
+// GetStampIssuer finds a stamp issuer by batch ID.
+func (ps *Service) GetStampIssuer(batchID []byte) (*StampIssuer, error) {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+	for _, st := range ps.issuers {
+		if bytes.Equal(batchID, st.batchID) {
+			return st, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+// Load loads all active batches (stamp issuers) from the statestore.
+func (ps *Service) Load() error {
+	n := 0
+	if err := ps.store.Iterate(ps.key(), func(key, _ []byte) (stop bool, err error) {
+		n++
+		return false, nil
+	}); err != nil {
+		return err
+	}
+	for i := 0; i < n; i++ {
+		st := &StampIssuer{}
+		err := ps.store.Get(ps.key(i), st)
```

Comment:
yes, how to do that?

on the snapshot stufff, we need to take care of reorgs and/or confirmation time of n blocks 
we need  @paxthemax and @ralph-pichler to figure that out

Reply:
Since all values are available in the iterator function above, it is possible to unmarshal them there and add to the Service. If Load and Save are not called concurrently, this is not of such importance, and can be left as it is.

Reply:
state load and save never called concurrently indeed